# Experiment a30cce27570af6fe1d637fbb2ff9fd640c954497cf34cfc9e8cbaa62454439a7
# Base 4a603c669103afff6e2ff8ff6ab2edc399b3cb42
diff --git a/src/Dib.cpp b/src/Dib.cpp
index 42905f34a..69d610b59 100644
--- a/src/Dib.cpp
+++ b/src/Dib.cpp
@@ -1257,16 +1257,17 @@ static void __cdecl xlatClut8(unsigned char* pb, unsigned long dwSize,
     } while (dwSize != 0);
 }
 
-static void __cdecl xlatClut4(unsigned char *buf, unsigned long count,
-    unsigned char *clut)
+static void __cdecl xlatClut4(unsigned char *pb, unsigned long dwSize,
+    unsigned char *xlat)
 {
-    while (count) {
-        unsigned int v = *buf;
-        unsigned int h = v >> 4;
-        *buf = (unsigned char)(clut[h] << 4) | clut[v & 0x0F];
-        ++buf;
-        --count;
-    }
+    if (dwSize <= 0)
+        return;
+
+    do {
+        *pb = xlat[*pb & 0x0f] | (xlat[*pb >> 4] << 4);
+        ++pb;
+        --dwSize;
+    } while (dwSize != 0);
 }
 
 static void __cdecl xlatRle8(unsigned char* data, unsigned long unused, unsigned char* xlat)
